pkg/oc/admin/migrate/legacyhpa/hpa.go

```diff
@@ -0,0 +1,209 @@
+package legacyhpa
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	autoscaling "k8s.io/kubernetes/pkg/apis/autoscaling"
+	autoscalingclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/autoscaling/internalversion"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+
+	"github.com/openshift/origin/pkg/oc/admin/migrate"
+	"github.com/openshift/origin/pkg/oc/cli/util/clientcmd"
+)
+
+var (
+	defaultMigrations = map[versionKind]versionKind{
+		// legacy oapi group
+		{"v1", "DeploymentConfig"}: {"apps.openshift.io/v1", "DeploymentConfig"},
+
+		// webconsole shenaniganry
+		{"extensions/v1beta1", "DeploymentConfig"}:      {"apps.openshift.io/v1", "DeploymentConfig"},
+		{"extensions/v1beta1", "Deployment"}:            {"apps/v1", "Deployment"},
+		{"extensions/v1beta1", "ReplicaSet"}:            {"apps/v1", "ReplicaSet"},
+		{"extensions/v1beta1", "ReplicationController"}: {"v1", "ReplicationController"},
+	}
+
+	internalMigrateLegacyHPALong = templates.LongDesc(fmt.Sprintf(`
+		Migrate Horizontal Pod Autoscalers to refer to new API groups
+
+		This command locates and updates every Horizontal Pod Autoscaler which refers to a particular
+		group-version-kind to refer to some other, equivalent group-version-kind.
+
+		The following transformations will occur:
+
+%s`, prettyPrintMigrations(defaultMigrations)))
+
+	internalMigrateLegacyHPAExample = templates.Examples(`
+		# Perform a dry-run of updating all objects
+	  %[1]s
+
+	  # To actually perform the update, the confirm flag must be appended
+	  %[1]s --confirm
+
+	  # Migrate a specific group-version-kind to the latest preferred version
+	  %[1]s --initial=extensions/v1beta1.ReplicaSet --confirm
+
+	  # Migrate a specific group-version-kind to a specific group-version-kind
+	  %[1]s --initial=v1.DeploymentConfig --final=apps.openshift.io/v1.DeploymentConfig --confirm`)
+)
+
+func prettyPrintMigrations(versionKinds map[versionKind]versionKind) string {
```

Comment:
This needs to do sorting so the output is stable.

Reply:
done